### PR TITLE
feature/products-device-list

### PR DESCRIPTION
**Pull request by busticated**
## Description

Adds `particle product` namespace and subcommand to list product devices ([ch](
https://app.clubhouse.io/particle/story/44999/enable-listing-product-devices))

## How to Test

_Note: requires you to have a Particle product with one or more devices_

Run `particle product device list --help` and explore the command and options.

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [x] Docs have been updated
- [x] CI is passing

### CODE REVIEW

File: test/e2e/product.e2e.js [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+		// TODO (mirande): device's `product_id` field should probably be a string
+		it('Lists devices using the `--json` flag', async () => {
+			const args = ['product', 'device', 'list', PRODUCT_01_ID, '--json'];
+			const { stdout, stderr, exitCode } = await cli.run(args);
+			const json = parseAndSortDeviceList(stdout);
+
+			expect(json).to.have.all.keys('meta', 'data');
+			expect(json.meta).to.have.all.keys('version', 'current', 'next', 'previous');
+			expect(json.meta.version).to.equal('1.0.0');
+			expect(json.data).to.have.lengthOf(2);
```
Comment (busticated) [inline test/e2e/product.e2e.js]: `at.least(2)`

File: test/e2e/product.e2e.js [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+			expect(json).to.have.all.keys('meta', 'error');
+			expect(json.meta).to.have.all.keys('version');
+			expect(json.meta.version).to.equal('1.0.0');
+			expect(json.error).to.have.property('message').that.is.a('string');
+			expect(json.error.message).include('HTTP error 404');
+			expect(stderr).to.equal('');
+			expect(exitCode).to.equal(1);
+		});
+
+		it('Fails to show device detail when `deivice` is unknown', async () => {
```
Comment (busticated) [inline test/e2e/product.e2e.js]: typo: `deivice`

File: test/e2e/product.e2e.js (hunk 1 above)
Comment (busticated) [reply test/e2e/product.e2e.js]: ✅

File: test/e2e/product.e2e.js (hunk 2 above)
Comment (busticated) [reply test/e2e/product.e2e.js]: ✅

**Comment by monkbroc** [comment]
One observation is that the list does not filter the quarantined devices. The device sam-photon1 is quarantined. My suggestion for handling this is to filter out all devices with `quarantined: true` by default. When passed the `--quarantined` flag, only output the quarantined devices. The API will return all quarantined devices on the first page.

```
$ particle product device list 8177 -g us-mi-julien
sam-photon1 [3d0034000c47363330353437] (Product 8177) is offline
p001 [430025000547363339343638] (Product 8177) is offline
p004 [2b0026000a47363339343638] (Product 8177) is offline
p003 [230030000a47363339343638] (Product 8177) is offline
p002 [3e003b000c47363433353735] (Product 8177) is offline
p005 [3b001b000347363339343638] (Product 8177) is offline
```

File: src/app/command-processor.js [hunk 3]
```diff
@@ -615,7 +615,7 @@ function createCommand(category, name, description, options){
  */
 function parse(command, args){
 	Yargs.reset();
-	Yargs.wrap(Yargs.terminalWidth());
+	Yargs.wrap(null);
```
Comment (monkbroc) [inline src/app/command-processor.js]: Is this change on purpose? I don't have strong feelings about wrapping one way or another so you own the final decision. This is the difference for running `particle help` with and without wrapping.

image

image

File: src/lib/ui/index.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+			const connectedState = device.connected ? 'online' : 'offline';
+			let name;
+
+			if (!device.name || device.name === 'null'){
+				name = '<no name>';
+			} else {
+				name = device.name;
+			}
+
+			if (device.connected){
```
Comment (monkbroc) [inline src/lib/ui/index.js]: Regretably the user device API endpoint returns `device.connected` and the product device API endpoint returns `device.online` when a device is online :cry: 
So you have to check both to report if a device is online.

File: src/app/command-processor.js (hunk 3 above)
Comment (busticated) [reply src/app/command-processor.js]: >Is this change on purpose?

yes. i was trying to fix a bug with the param type labels (e.g. `[boolean]', etc) - see the full diff over here 👉 https://github.com/particle-iot/particle-cli/pull/544/commits/bb4936cfc42009f458935f3b5cbb1b5d4a5190dc

>This is the difference for running particle help with and without wrapping.

we're a bit wider than the ideal of 80cols but things should look normal if you expand the terminal window - for that help screen:

```sh
$tput cols
125
```

ultimately, the best thing would be to fix the buggy type label handling without messing with wrapping, etc. i wasn't able to figure out how to thread that needle - all pointers welcome.

in the meantime, i'll highlight this change with the team and get feedback :+1:

Comment (wraithan) [reply src/app/command-processor.js]: I think this makes the output consistent for testing, maybe we want to be able to turn this on for testing but not for production.

Comment (busticated) [reply src/app/command-processor.js]: >maybe we want to be able to turn it off for testing

not a bad suggestion :+1:

File: src/lib/ui/index.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
+const os = require('os');
+const Chalk = require('chalk').constructor;
+const platformsById = require('../../cmd/constants').platformsById;
+
+
+module.exports = class UI {
```
Comment (wraithan) [inline src/lib/ui/index.js]: hype

**Comment by busticated** [comment]
> One observation is that the list does not filter the quarantined devices. The device sam-photon1 is quarantined. My suggestion for handling this is to filter out all devices with `quarantined: true` by default. When passed the `--quarantined` flag, only output the quarantined devices. The API will return all quarantined devices on the first page.
> 
> ```
> $ particle product device list 8177 -g us-mi-julien
> sam-photon1 [3d0034000c47363330353437] (Product 8177) is offline
> p001 [430025000547363339343638] (Product 8177) is offline
> p004 [2b0026000a47363339343638] (Product 8177) is offline
> p003 [230030000a47363339343638] (Product 8177) is offline
> p002 [3e003b000c47363433353735] (Product 8177) is offline
> p005 [3b001b000347363339343638] (Product 8177) is offline
> ```

@monkbroc does the API support filtering by `quarantined`? if not, i'd be concerned about handling the case where a product has more quarantined devices than the page size and returning nothing. might be easier to discuss in person :+1:

File: src/lib/ui/index.js (hunk 4 above)
Comment (busticated) [reply src/lib/ui/index.js]: fixed over here 👉 https://github.com/particle-iot/particle-cli/pull/544/commits/9006100a68a1991e9a1def1fd509995631e09080

File: src/app/command-processor.js (hunk 3 above)
Comment (busticated) [reply src/app/command-processor.js]: ✅ done in https://github.com/particle-iot/particle-cli/pull/544/commits/5f92d8adce427a8a5f3b4d956296dcdef8fbe807
